3257 Fix long snippets in recap results and removed safe tag from plain_text

albertisfu opened the pull request:
This PR solves two small issues described in #3257

The first issue is related to long `plain_text` snippets. The problem arose because, when `plain_text` is not highlighted, no fragments are created for the snippet, resulting in the display of the entire field content. 

To resolve this, I added the `truncatewords` template tag and set it to 50 words, ensuring that only a couple of lines are displayed. I applied this tweak to the description field as well, which can also be lengthy.

Now `plain_text` is truncated.
<img width="672" alt="Screenshot 2023-10-24 at 13 39 58" src="https://github.com/freelawproject/courtlistener/assets/486004/a09bbeda-5980-4fa5-a5a8-3b4e3752b05f">


The other issue solved in this PR is the one related to the bold tag within the `plain_text`.

 
<img width="775" alt="Screenshot 2023-10-24 at 13 34 43" src="https://github.com/freelawproject/courtlistener/assets/486004/9ff2ba36-7a56-4155-9b0b-bd0afe9755d0">

The problem was that the `plain_text` contained a "bold" tag which was interpreted as HTML due to the `safe` tag.

```
> [21 U.S.c. §§ 841(a)(1>
) and (b) <3.> <B>§
```

So, I removed the safe tag from the `plain_text` field. This should be fine if we don't expect HTML to be extracted from PDFs and rendered as HTML, is that correct?

mlissner commented:
Hm, I'm a bit concerned that pulling these documents from the index is bad for performance though. Is there a way not to pull so much text from the index if we don't have them highlighted? 

Also, for the descriptions could we do a JS thing to show a little `[+]` icon to allow viewing the full description? We do that currently on parties, I believe?

albertisfu commented:
Sure, I found out that in fact, it's possible to only return a fragment of the field in case no terms are highlighted, using the setting `no_match_size` in the highlights, I created the setting: `NO_MATCH_HL_SIZE` and set it to 500 chars.

So now if a field that should be highlighted in fragments like the `plain_text` doesn't match any terms during the process, the first 500 chars of the field are displayed.

- Also, the whole field is excluded from the ES response, in this case, `plain_text`,  to improve the performance.

- I removed truncation to the description field and instead I applied the `read_more` template tag set to 100 words.

So now a child document looks like this:

No highlighted terms:
<img width="734" alt="Screenshot 2023-10-24 at 19 16 53" src="https://github.com/freelawproject/courtlistener/assets/486004/72838815-4db3-4168-afb4-de5b4f650d57">

Highlighted terms:
<img width="680" alt="Screenshot 2023-10-24 at 19 26 21" src="https://github.com/freelawproject/courtlistener/assets/486004/93a69e8b-3d22-4c80-aebe-5a57ffdab043">

mlissner reviewed (approved):
Looks fine to me. @ERosendo, once you're happy, let's merge.

mlissner commented:
Thanks for all those answers and for finding the better solution. I think this will help with performance too, which is great.

ERosendo commented on cl/lib/elasticsearch_utils.py:
```diff
@@ -584,13 +584,26 @@ def build_has_child_query(
 
+    fields_to_exclude = []
     for field, fragment_size in highlighting_fields.items():
         number_of_fragments = 0
+        no_match_size = 0
         if fragment_size:
             number_of_fragments = 1
+            # In fields that have a defined fragment size in their HL mapping
+            # e.g., SEARCH_RECAP_CHILD_HL_FIELDS, a 'no_match_size' parameter
+            # is also set. If there are no matching fragments to highlight,
+            # this setting will return a specified amount of text from the
+            # beginning of the field.
+            no_match_size = settings.NO_MATCH_HL_SIZE
+            if not field.endswith("exact"):
+                # The original field is excluded from the response to avoid
+                # returning the entire field from the index.
+                fields_to_exclude.append(field)
```
We should refactor this code using conditional expressions to avoid the nested if-statement. For example:

```python
number_of_fragments = 1 if fragment_size else 0
no_match_size = settings.NO_MATCH_HL_SIZE if fragment_size else 0

if fragment_size and not field.endswith("exact"):
    fields_to_exclude.append(field)
```

albertisfu replied:
Great,  yeah this makes the code simpler, I've applied the suggestion!

albertisfu commented:
@ERosendo  Thank you. I've implemented your suggestions in the code, including those for the frontend that we discussed on Slack.

- Now, we no longer display the ellipsis at the beginning of the `plain_text` snippet when there is no highlighted content. In such cases, the first lines are already displayed, so showing an initial ellipsis is unnecessary.
![Screenshot 2023-10-25 at 12 11 17](https://github.com/freelawproject/courtlistener/assets/486004/75843859-1c4a-48bc-9c3b-fe84336a050a)

- The ellipsis at the beginning will only appear when there is highlighted content in the snippet.
![Screenshot 2023-10-25 at 12 10 57](https://github.com/freelawproject/courtlistener/assets/486004/ab18fc87-8a52-4df8-a9aa-c1d990d1ab85)

- Additionally, I've added a top margin to `plain_text` as you suggested, to better differentiate it from the description, it can be noticed in the previous screenshots.
